metrics/tags.go

```diff
@@ -3,12 +3,162 @@ package metrics
 import (
 	"bytes"
 	"encoding/json"
+	"errors"
 	"sort"
 	"strings"
+
+	"github.com/mailru/easyjson"
+	"github.com/mailru/easyjson/jlexer"
+	"github.com/mailru/easyjson/jwriter"
+	"github.com/mstoykov/atlas"
 )
 
+// A TagSet represents an immutable set of metric tags. For the efficient and
+// thread-safe storage of the key=value tag pairs, it uses the
+// https://github.com/mstoykov/atlas data structure.
+//
+// Assuming all tag sets start from the same root (see Registry.RootTagSet()),
+// you can compare *TagSet values of different metric Samples with the `==` to
+// check if they have the same tags, and you can also use *TagSet values for map
+// indexes and caching.
+//
+// See also the TimeSeries type for comparing a Sample's {metric+tags} for
+// equality at the same time.
+type TagSet atlas.Node
+
+// With returns another TagSet object that contains the combination of the
+// current receiver tags and the name=value tag from its parameters.
+//
+// It doesn't modify the receiver, it will either return an already existing
+// TagSet with these tags, if it exists, or create a new TagSet with them and
+// return it.
+//
+// If a tag with the specified name already exists in the set, it will be
+// overwritten with the new value in the returned set.
+func (ts *TagSet) With(name, value string) *TagSet {
```

Comment:
nitpick, feel free to ignore 🐻 

When looking at it in a context where it's used, I found `With` and `Without` to be kind of abstract. Although I see where it comes from, as those are immutable methods, and we need to carry on the meaning that a new instance will be created out of it,  I wondered if `Including`, and `Excluding` could be more explicit and lead to "sentence-like" when taken in their calling context:
- `myTag := myRegistry.TagSet().Including("a", "1")` instead of `myRegistry.RootTagSet().With("a", "1")`
- `tags := currentTags.Excluding("key")`

Reply:
Hmm, I think I personally slightly prefer `With()` and `Without()`, but I am fine with either, so I can rename if everyone else prefers `Including()` and `Excluding()` :thinking: 

I prefer `With()` and `Without()` because they are shorter and because I somewhat associate `Including` and `Excluding` with filtering more than I associate them with sets. But again, both will work, so @mstoykov @codebien @olegbespalov , please comment if you'd like me to rename them.

Reply:
And another reason, consider the following pieces of code:
```go
tags := tagSet().With("foo", "bar").With("foo", "baz")
```
and
```go
tags := tagSet().Including("foo", "bar").Including("foo", "baz")
``` 

In the first example, while it's not clear that the second `With()` statement will overwrite the first (which it will), it's not suggesting the opposite either. But the second example might mislead a reader to think that the final `tags` value will include both `foo` tags, which it won't.

Reply:
Counter suggestion: `Plus`/`Minus` 

`tags := tags.Plus("foo" , "bar")` 

`tags := tagSet().Plus("foo", "bar").Plus("foo", "baz")`

same number of letters but I would say the same probem as `Including` - not really communicating that the same tags will be overwritten.

I am not certain that `With` does this any better though :shrug:. 

I did mention before that `With` and `Without` seem very similar while scanning code thouhg so :shrug: .

Also `And`/`Not` might be fine ... although we lean really into the `Set` part at this point:

`tags := tags.And("foo", "bar").And("foo", "baz").Not("bar")`

Not really certain it  communicates anything better ... but even less symbols :P

Reply:
hmm `With()` doesn't communicate straight addition / accumulation, which both `Including` and `Plus` do. In my mind, I read `tags := rootTagSet().With("scenario", "foo")` as "tags is the root tag set **with** a `scenario` tag equal to `foo`". Whether the root tag set had such a tag or not doesn't matter, "with" suggest a singular tag value. Whereas `Plus`, `Including` and `And` kind of suggest both will be kept. 

The only other word that I can think of, similar to `With` is `Containing`, but it's longer, it doesn't flow as well when I read it in my head and it doesn't have a nice opposite word :sweat_smile:

Reply:
I kinda like the `New*` approach you propose @olegbespalov, it does make the copy and the immutability explicit. The only concern I would have is that the `New` pattern in Go is pretty much established as a way to instantiate an object from scratch, rather than as an already existing's instance method ❓

Reply:
Yeah, `NewWith()`, `NewWithout()`, etc. would be both more verbose and somewhat confusing with regular constructors :thinking: After all, the majority of time `With()` won't return a new object, but an already existing atlas `Node`... I am not completely against it, but I don't think it's an improvement.

@mstoykov, @oleiade, can you restate what specifically you find objectionable about 
```go
func (ts *TagSet) With(name, value string) *TagSet { /* ... */ }
func (ts *TagSet) Without(name string) *TagSet { /* ... */ }
```

Give some example where their usage is confusing or something like that? I am somewhat at a loss exactly what their problems are that we are trying to solve :sweat_smile: Having re-written the PR with them, the code seems perfectly readable and understandable to me, but I might be missing something. And in the interests of avoiding bikeshedding, I'd like to reach some consensus here soon :bow:

Reply:
For sure, from scratch (an `existing` as an argument) and `New*` is a standalone method, but I provided it just as my consideration (FYI); 

In any case for code below is more readable with the existing chain approach instead than with `New*`:

Good: 
```
x.With("tag", "value").With("tag2", "value").With("tag2", "value")
```

vs.

Bad:
```
New(New(New(x, "tag", "value"), "tag2", "value"), "tag3", "value")
```

Reply:
:-1: for `New*` for the reasons @na-- gave more or less. 

@na-- My original issue was `With` and `Without` being way too similar when just scrolling through. Nobody else has mentioned this and I very much doubt that we will use this *everywhere* so it's unlikely to be a problem.

More or less my suggestions were about trying to figure out a slightly better pair of names, but I also think `With`/`Without` are currently the best so far. Especially as the others usually have way bigger downsides. 

I personally also prefer if we leave this discussion and if there is such a big problem later on we can rename (and leave `With`/`Without` for a version as deprecated). I would expect that if we would want to rename them we will have more concrete problems and arguments.

Reply:
Alright, then it seems like there is at least a tentative majority (if not an outright consensus) that `With()` and `Without()` are good enough, or at least the least bad alternative... :sweat_smile:  I'll resolve this conversation for now then. 

And yeah, we can always rename this in the future, if we find some big issues or figure out something substantially better.

Reply:
I like `With`, it is explicit enough for the purpose, IMHO. I would like to not have a similar word for the negative form so we could mix the proposals. 

So we could have `With` and `Drop` - or `Remove` or `Delete` ( eventually `Del` in short).

Reply:
I think I like `Drop` the most :thinking: and it seems I finally have someone who also things they are too similar :tada:. 

But again I do think we can change this a bit later again. We certainly do have time until this is released and we will likely add and change a bunch of stuff, so not certain if renaming now is that beneficial unless the majority of the time decides it's really better.

As at this point it still seems like different people have different ideas on how this should look and `With`/`Without` seems to still be the okay+ for most people. So ... uh please write your opinion if you disagree :P

Reply:
I disagree, we are using `Without` in only one place, so it seems quite quick to change. We have some time before we are going to merge this so I think it's better to change it now if we have an agreement on one of the previous.

Instead, if we will continue to have different opinions let's keep it as-is.

Reply:
I am for keeping them as they are. You can't mix them, because `With()` and `Without()` have a different number of arguments. And besides, after https://github.com/grafana/k6/pull/2654 is merged, almost nobody should directly use `Without()` outside of tests, it will probably only be used in these 2 places in the code:
https://github.com/grafana/k6/blob/6ed8668af8eba6669e79d141f8e3ec942fa8e367/metrics/tags.go#L185-L186
https://github.com/grafana/k6/blob/6ed8668af8eba6669e79d141f8e3ec942fa8e367/metrics/tags.go#L214-L216